examples/vision/siamesenetwork.py

```diff
@@ -0,0 +1,398 @@
+)
+
+siamese_network = Model(
+    inputs=[anchor_input, positive_input, negative_input], outputs=distances,
+)
+
+"""
+# Putting everything together
+
+We now need to implement a model with custom training loop so we can compute
+the Triplet Loss using the three embeddings produced by the Siamese network.
+
+Here is the definition of Triplet Loss implemented on this example:
+`L(A,P,N) = max(||f(A)-f(P)||**2 - ||f(A)-f(N)||**2 + alpha, 0)`
+"""
+
+"""
+Let's create a `Mean` metric instance to track the loss of the training process.
+"""
+
+loss_tracker = metrics.Mean(name="loss")
+
+
+class SiameseModel(Model):
+    """
+    Model implementing a custom training loop to compute the Triplet Loss
+    using the three embeddings produced by the Siamese network.
+
+    Here is the definition of Triplet Loss:
+        L(A,P,N) = max(||f(A)-f(P)||**2 - ||f(A)-f(N)||**2 + alpha, 0)
+
+    """
+
+    def __init__(self, siamese_network, alpha=0.5):
+        super(SiameseModel, self).__init__()
+        self.siamese_network = siamese_network
+        self.alpha = alpha
+
+    def call(self, inputs):
+        self.siamese_network(inputs)
+
+    def train_step(self, data):
+        # GradientTape is a context manager that records every operation that you do inside.
+        # We are using it here to compute the loss so we can get the gradients and apply
+        # them using the optimizer specified in `compile()`.
+        with tf.GradientTape() as tape:
+            anchor, positive, negative = data
+
+            # The output of the network is a tuple containing the distances
+            # between the anchor and the positive example, and the anchor and
+            # the negative example.
+            ap_distance, an_distance = self.siamese_network(
+                (anchor, positive, negative)
+            )
+
+            # Computing the Triplet Loss by subtracting both distances and
+            # making sure we don't get a negative value.
+            loss = ap_distance - an_distance
+            loss = tf.maximum(loss + self.alpha, 0.0)
+
+        # Let's get the gradients (loss with respect to trainable weights)
+        gradients = tape.gradient(loss, self.siamese_network.trainable_weights)
+
+        # Applying the gradients on the model using the specified optimizer
+        self.optimizer.apply_gradients(
+            zip(gradients, self.siamese_network.trainable_weights)
+        )
+
+        # Let's update and return the loss metric.
+        loss_tracker.update_state(loss)
+        return {"loss": loss_tracker.result()}
+
+    @property
+    def metrics(self):
+        # We need to list our metric here so the `reset_states()` can be
+        # called automatically.
+        return [loss_tracker]
+
+
+"""
+# Training
+
+We are now ready to train our model.
+"""
+
+siamese_model = SiameseModel(siamese_network)
+siamese_model.compile(optimizer=optimizers.Adam(0.0001))
+siamese_model.fit(dataset, epochs=20)
```

Comment:
We should also validate the training on a `validation_data` dataset here, to monitor overfitting.

Reply:
@hazemessamm Do you want to tackle this one? I already split the dataset into `train_dataset` and `val_dataset.` Let me know if you need help with this, and I'm happy to jump back into it.

Reply:
@svpino okay great I will finish this part

Reply:
@svpino I assume that in order to validate the training I need to override the ```test_step()``` right ?

Reply:
Yup, we should be able to do it by overwriting `test_step()` and computing a new `val_loss` metric just like we are computing `loss.`

Our goal is to display the `val_loss` alongside `loss` to determine whether we are overfitting.

Reply:
@svpino I did that but the problem is that it does not appear after the epoch ends and I tried it on a toy example and it worked so I think there is a problem with the validation_dataset

Reply:
@svpino I will push the code so you can check it

Reply:
Alright, this should be good to go now.

We have a custom `test_step()` that computes the `val_loss.` I extracted the common functionality from `train_step()` and `test_step()` into a separate function.